mailpoet/mailpoet
# Add support for the invisible ReCaptcha [MAILPOET-4145]

This PR adds support for the invisible ReCaptcha in the subscribe forms. Before, we only supported the checkbox ReCaptcha. The API is almost identical, but there are some differences. Because of those differences, it was necessary to change the code that handles the form submission on public.jsx a bit. 

For the invisible version, we need to call recaptcha.execute(), and then the ReCaptcha API calls the pre-defined callback set inside renderCaptcha() with a token if the captcha was solved successfully. The callback then calls submitSubscribeForm() with the token. For the checkbox version, we get the token synchronously inside the code that handles the form submission when recaptcha.getResponse() is called. So we can call submitSubscribeForm() directly.

I would suggest reviewing this PR commit by commit. The first commit makes the necessary changes to the admin to add a new setting to use the invisible ReCaptcha. The second commit just makes some organizational changes to public.jsx that are required for the next commit due to the no-use-before-define ESLint rule. Finally, the third commit changes the frontend code to display the invisible ReCaptcha if this is the selected setting.

**Testing instructions**

- Test that **submitting a form on the front-end works** for the three different types of captcha that we support (Built-in, Google reCAPTCHA v2 Checkbox, Google reCAPTCHA v2 Invisible) and also **when the captcha is disabled**.
- You will need to enter `Google ReCaptcha Keys` in `Settings` > `Advanced tab`... and find your keys in the `P2` doc
- Ensure that the **new setting in the admin works** according to the specs in the Jira ticket.
- It is a bit tricky to test that the invisible ReCaptcha is working as most of the time, it works without interfering with the form submission flow. You add a breakpoint to SubscriberSubscribeController::validateCaptcha() and confirm that this method is getting a ReCaptcha token and that the request that it makes to the ReCaptcha server has a successful response.
- Links to the ReCaptcha API documentation in the Jira ticket.

[MAILPOET-4145]

[MAILPOET-4145]: https://mailpoet.atlassian.net/browse/MAILPOET-4145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
